common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.function.Function;
+
+import static io.netty.util.internal.PromiseNotificationUtil.tryFailure;
+
+/**
+ * Combinator operations on {@linkplain Future futures}.
+ *
+ * @implNote The operations themselves are implemented as static inner classes instead of lambdas to aid debugging.
+ */
+public final class Futures {
+    private static final InternalLogger LOGGER = InternalLoggerFactory.getInstance(Futures.class);
+    private static final PassThrough<?> PASS_THROUGH_CONSTANT = new PassThrough<Object>();
+
+    /**
+     * Creates a new {@link Future} that will complete with the result of the given {@link Future} mapped through the
+     * given mapper function.
+     * <p>
+     * If the given future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but the given future will be unaffected.
+     *
+     * @param future The future whose result will flow to the returned future, through the mapping function.
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R>    The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    public static <V, R> Future<R> map(Future<V> future, Function<V, R> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+        if (future.isSuccess()) {
+            return future.executor().submit(new CallableMapper<>(future, mapper));
+        }
+        Promise<R> promise = future.executor().newPromise();
+        future.addListener(new Mapper<>(promise, mapper));
+        promise.addListener(future, PromiseNotifier::propagateCancel);
+        return promise;
+    }
+
+    /**
+     * Creates a new {@link Future} that will complete with the result of the given {@link Future} flat-mapped through
+     * the given mapper function.
+     * <p>
+     * The "flat" in "flat-map" means the given mapper function produces a result that itself is a future-of-R, yet this
+     * method also returns a future-of-R, rather than a future-of-future-of-R. In other words, if the same mapper
+     * function was used with the {@link #map(Future, Function)} method, you would get back a {@code Future<Future<R>>}.
+     * These nested futures are "flattened" into a {@code Future<R>} by this method. Note that the future returned by
```

Comment:
I feel we can do without this note and the reason. The `@return` mentions that a new `Future` is returned.